### DISCUSSION TITLE
May Cake load pre-installed addin as Package?

### DISCUSSION THREAD

**Discussion by ymartin59** [cake-build/cake#3276] [Q&A] [answered]
Hello

I have discovered that most Cake addin may be pre-installed on system or user scope with "Install-Package" command from https://github.com/SharpeRAD/Cake.Powershell

But whatever I use as "#addin" expression, either short "Cake.Powershell" or long "#addin nuget:?package=Cake.Powershell&version=1.0.0", Cake always (probably download) and create "Tools/AddIn/Cake.Powershell.1.0.0"

Is there a way to enforce Cake not to download-install in Tools/AddIn when "Get-Package Cake.Powershell" is available?

I am preparing a Docker windows container image with tools to then build a project sources without any internet connection... so the pre-installed package would help in that context (and avoid to store tools in git repository)

Have you got any hint how to achieve this?

**Comment by augustoproiete** [comment]
Have you tried adding the package to the `Tools/AddIn` folder exactly as Cake would do? That might do the trick.

**Reply by ymartin59** [reply]
I agree it should work. But I am looking for a more efficient way for my CI automation than implementing a pre-build copy of system-wide installed packages into my just-checked-out Git working copy.

I was wondering if `using Cake.Powershell` would work without `#addin` directive or else if locating `Cake.Powershell.dll` into `C:\ProgramData\chocolatey\lib\cake.portable\` would be enough for Cake to gain access to functions.

If it makes sense, why not extend `#addin` directive to support `packagemanagement:?package=Cake.Powershell&version=1.0.0` expression (for instance) and leveraging PackageManagement support in powershell cmdlets Find-Package/Install-Package/Get-Package ?

**Reply by augustoproiete** [reply]
I'd be interested in understanding more of the scenario, such as:

- What are the system-wide packages?

- Where do they come from?

- How are they being installed in the image? 

- Do they need to be system-wide? Why?

- Would having a `setup.cake` file containing all the addins that you need to pre-download, work? You'd just have to execute `dotnet cake setup.cake` instead of executing the process that you have today to download packages

- **More importantly**: Do you have to write *any* PowerShell at all? Is there anything preventing you from using Cake from start to finish?

---

Now to your question on extending preprocessor directive to support something like `packagemanagement:?package=...` it's something you could achieve today by creating a [Cake module](https://cakebuild.net/docs/extending/modules/) that handles this `packagemanagement` scheme and locates the assembly in the place you want.

You can look at the [code of the many different modules](https://github.com/cake-contrib?q=Module&type=&language=&sort=) that can be used to extend Cake to get an idea of how you would build your own.

ps: Feel free to add the above as a feature request in the [Ideas area](https://github.com/cake-build/cake/discussions/categories/ideas) if you think it would be valuable to have this `packagemanagement` module as part of Cake in a future version.

**Reply by ymartin59** [reply]
Many thanks for your reply.

As README file at https://github.com/SharpeRAD/Cake.Powershell proposes to run `Install-Package Cake.Powershell` to get Powershell PackageManagement download and extract NuGet package to a "standard location" (either on system if admin or with `-Scope CurrentUser`), I was wondering if Cake would be able to grab it directly from that installed location, easy to retrieve from `Get-Package Cake.Powershell` output.

Why I expect to store all build tools including Cake addins in Docker image (so "system-wide" and outside Git working copy)? Because it will then run build targets WITHOUT internet access.

And I also expect my Cake target to run "the same way" on a the "offline" Docker image or on a developer regular workstation.

And yes, I have to invoke complex Powershell tooling to pack Dynamics CRM v8 (on-premise) Solution with XRM.CI.Framework according to https://dynamics-chronicles.com/article/xrm-ci-framework-powershell-scripts-automate-dynamics-365-deployment - no question to build on cloud / azure devops for my project...

**Comment by augustoproiete** [comment, marked as answer]
Thanks for the context.

AFAIK the [`Install-Package` you see on the README of `Cake.Powershell`](https://github.com/SharpeRAD/Cake.Powershell#referencing) is meant to be run from within Visual Studio, in the Package Manager Console, so that you can use that Cake Addin in [Cake Frosting projects](https://cakebuild.net/docs/running-builds/runners/cake-frosting) (as opposed to Cake Scripting which uses the `#addin` directive).

Regardless of that, I believe we're back to my very [initial suggestion](https://github.com/cake-build/cake/discussions/3276#discussioncomment-475205):
> _add the package(s) to the Tools/AddIn folder exactly as Cake would do_

**The Tools/Addin folder doesn't have to reside inside the Git working folder**. It can be anywhere in the machine. You just have to tell Cake where it is, either by setting a `CAKE_PATHS_ADDINS` environment variable, or by providing the `--paths_addins` argument, or by dropping a [`cake.config`](https://github.com/cake-build/resources/blob/master/cake.config#L9-L12) file alongside your `.cake` script.

More details: https://cakebuild.net/docs/running-builds/configuration/default-configuration-values#addins-path

So you could have a `setup.cake` that you run to pre-download the addins to a folder that can be stored anywhere in the container:

```
#addin "nuget:?package=Cake.Powershell&version=1.0.0&loaddependencies"
#addin "nuget:?package=Cake.AnotherAddin&version=1.0.0"
// etc...
```

And in your container you set the `CAKE_PATHS_ADDINS` environment variable to the folder where you pre-stored stored the addins.

**Reply by ymartin59** [reply]
OK thanks. I will give a try with `CAKE_PATH_ADDINS`.
